tableauserverclient/models/database_item.py

```diff
@@ -0,0 +1,241 @@
+import xml.etree.ElementTree as ET
+
+from .permissions_item import Permission
+
+from .property_decorators import property_is_enum, property_not_empty, property_is_boolean
+from .exceptions import UnpopulatedPropertyError
+
+
+class DatabaseItem(object):
+    class ContentPermissions:
+        LockedToProject = 'LockedToDatabase'
+        ManagedByOwner = 'ManagedByOwner'
+
+    def __init__(self, name, description=None, content_permissions=None):
+        self._content_permissions = None
+        self._id = None
+        self.description = description
+        self.name = name
+        self.content_permissions = content_permissions
+
+        self._contact_id = None
+        self._certified = None
+        self._connector_url = None
+        self._permissions = None
+        self._certification_note = None
+
+        self._tables = None
+        self._default_table_permissions = None
+
+    @property
+    def content_permissions(self):
+        return self._content_permissions
+
+    @property
+    def permissions(self):
+        if self._permissions is None:
+            error = "Project item must be populated with permissions first."
+            raise UnpopulatedPropertyError(error)
+        return self._permissions()
+
+    @property
+    def default_table_permissions(self):
+        if self._default_table_permissions is None:
+            error = "Project item must be populated with permissions first."
+            raise UnpopulatedPropertyError(error)
+        return self._default_table_permissions()
+
+    @content_permissions.setter
+    @property_is_enum(ContentPermissions)
+    def content_permissions(self, value):
+        self._content_permissions = value
+
+    @property
+    def id(self):
+        return self._id
+
+    @property
+    def name(self):
+        return self._name
+
+    @name.setter
+    @property_not_empty
+    def name(self, value):
+        self._name = value
+
+    @property
+    def description(self):
+        return self._description
+
+    @description.setter
+    def description(self, value):
+        self._description = value
+
+    @property
+    def embedded(self):
+        return self._embedded
+
+    @property
+    def certified(self):
+        return self._certified
+
+    @certified.setter
+    @property_is_boolean
+    def certified(self, value):
+        self._certified = value
+
+    @property
+    def certification_note(self):
+        return self._certification_note
+
+    @certification_note.setter
+    def certification_note(self, value):
+        self._certification_note = value
+
+    @property
+    def metadata_type(self):
```

Comment:
This one and a few below are not defined in the constructor. Also for port, should it be an integer property?

Reply:
Not all of these properties will exist for all types (though `metadata_type` should).

Right now it lazily adds them based on whatever we pass to `_set_values`. 
We could enumerate them all, but I thought it was a lot of copy/paste work to add them there and to `_set_values`. 

Happy to do it if you think we should. It raises the question again about what makes it into constructor vs what doesn't.

Reply:
In my opinion, I think it's good practice to have all properties defined in the constructor. That way we can easily see all the properties that the class can have, and avoids the possibility of an error from accessing it before assigned from the response.
Looks like we've been keeping that pattern for the other models as well.